src/main/scala/ai/privado/languageEngine/java/tagger/source/SourceTagger.scala

```diff
@@ -0,0 +1,325 @@
+    val typeDeclMemberName = typeDeclMember.name
+    // Have started tagging Parameters as well, as in collection points sometimes there is no referencing Identifier present for a local
+
+    (cpgNodeCache.cachedIdentifier
+      .where(_.typeFullName(typeDeclVal))
+      .l ::: cpgNodeCache.cachedParameter.where(_.typeFullName(typeDeclVal)).l)
+      .foreach(impactedObject => {
+        if (impactedObject.tag.nameExact(Constants.id).l.isEmpty) {
+          storeForTag(builder, impactedObject, ruleCache)(
+            InternalTag.OBJECT_OF_SENSITIVE_CLASS_BY_MEMBER_NAME.toString,
+            taggerPart.ruleInfo.id
+          )
+          storeForTag(builder, impactedObject, ruleCache)(
+            Constants.id,
+            Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME
+          )
+          storeForTag(builder, impactedObject, ruleCache)(Constants.catLevelOne, CatLevelOne.DERIVED_SOURCES.name)
+        }
+        storeForTag(builder, impactedObject, ruleCache)(
+          Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME,
+          taggerPart.ruleInfo.id
+        )
+        // Tag for storing memberName in derived Objects -> user --> (email, password)
+        storeForTag(builder, impactedObject, ruleCache)(
+          taggerPart.ruleInfo.id + Constants.underScore + Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME,
+          typeDeclMemberName
+        )
+      })
+
+    // To Mark all field Access and getters
+    cpgNodeCache.tagAllFieldAccessAndGetters(builder, typeDeclVal, taggerPart.ruleInfo, typeDeclMemberName, true)
+  }
+}
+
+case class ExtendingDerivedSourceTaggerPart(typeDecl: TypeDecl, memberType: SecondLevelDerivedSourceTaggerPart)
+case class SecondLevelDerivedSourceTaggerPart(typeDecl: TypeDecl, ruleInfo: RuleInfo)
+case class DerivedSourceTaggerPart(typeDecl: TypeDecl, member: Member, ruleInfo: RuleInfo)
+case class CPGNodeCacheForSourceTagger(cpg: Cpg, ruleCache: RuleCache) {
+  val cachedIdentifier: List[Identifier] = cpg.identifier
+    .filterNot(item => item.name.equals(item.name.toUpperCase))
+    .filterNot(_.code.equals(Constants.thisConstant))
+    .l
+  val cachedParameter: List[MethodParameterIn] = cpg.parameter
+    .filterNot(item => item.name.equals(item.name.toUpperCase))
+    .filterNot(_.code.equals(Constants.thisConstant))
+    .l
+  val cachedMember = cpg.member.filterNot(item => item.name.equals(item.name.toUpperCase)).l
+  val cachedFieldAccess = cpg.method
+    .fullNameExact(Operators.fieldAccess, Operators.indirectFieldAccess)
+    .callIn
+    .dedup
+    .l
+  val derivedSourceTaggerParts = ruleCache.getRule.sources
+    .flatMap(rule => {
+      cachedMember
+        .name(rule.combinedRulePattern)
+        .map(memberNode => DerivedSourceTaggerPart(memberNode.typeDecl, memberNode, rule))
```

Comment:
There can be a further optimization, as after DirectSourceTagging, we know the member and the ruleId against which it was tagged

Reply:
It won't make much difference except the memory footprint. I was thinking about the approach you were suggesting if it would have increased the memory consumption by a large margin. As we are holding these caches in memory. 
However, after going through a test run, I didn't find memory consumption to be very high. 

refer this screenshot of CPU and memory usage - https://drive.google.com/file/d/1T5fupM51-t1IJqg16VcZLCYwU8T_GFXD/view?usp=sharing - (Usage before 1PM is for tagger steps. Post 1 PM starts with API Tagger and reachableByFlows.)
You can also refer to this granular matrix - https://docs.google.com/spreadsheets/d/1H7XOeRRwLRm-5CL74jYyo08CfygzgY8LUYwSfvqKv9o/edit?usp=sharing